### ISSUE TITLE
Waterview has issues parsing files that are parsed without problems by waterswap

### ISSUE THREAD

**Issue by sas103** [michellab/Sire#306]
Dear developers,

I am trying to run a waterview calculation on a system on which I've already run a regular waterswap calculation without any problems (in that case, I have specifically used the binary file repstruct_c0_BSA.rst as coordinates and 4ipe_2atp_2MG.solv_waterswap.top as topology, respectively).

When I now attempt to run a waterview calculation, either with the command

waterview -C config -c repstruct_c0_BSA.rst -t ../../preparation/4ipe_2atp_2MG.solv_waterswap.top -l atA

[binary coordinate file]

or with the command 

-C config -c repstruct_c0_BSA.rst7 -t ../../preparation/4ipe_2atp_2MG.solv_waterswap.top -l atA

[text coordinate file]

I get the two error messages and output present in the zip folder. I also get a slightly different error when trying to run the command with a "MDCRD" file.

Before I go on and try and tweak the code for myself, I was wondering if I might be missing something really obviously wrong wiht my input system, or if you have maybe encountered similar problems with waterview before.

Thanks in advance for your help, and please let me know whether I should post this message on the users' list instead.

Best regards,

Stefano

[files_to_reproduce_error.zip](https://github.com/michellab/Sire/files/4118225/files_to_reproduce_error.zip)

**Comment by chryswoods** [comment]
Hi Stefano. Thanks for raising this, and it is the best place. I’ll try to take a look this afternoon. I’m guessing there is some code rot as waterview is little used and not really marketted. The analytics show it has only ever been used twice in its history...!

Out of interest, what are you using waterview for, and how did you find it? Just curious as the only thing public about it is a talk I gave years ago and a YouTube video ;-)

**Comment by sas103** [comment]
Hi Chris, and many thanks for taking a look at this!

I know about waterview since back in the day when I was at the University of Bristol someone (probably you) must have mentioned it in one of their internal talks.  Now that I have gone back to using the "sire" suite for my current research, I remember that Adrian Mulholland occasionally mentioned doing (and I put it very approximately) "calculations to determine the energetic contribution of individual waters in the binding site". So I checked the siremol website, and I saw that there was this tab for waterview in the side menu (https://siremol.org/pages/apps/waterview.html).

Since I am studying the mechanism of an asymmetric dimeric ATPase, in which there are several (crystallographic) waters in the binding site that are crucial for hydrolysis, it makes sense to look at the energetics of binding. First with waterswap and now, I was thinking, also with waterview.

I hope this contextualises things a bit better. As I said, waterswap has worked perfectly on the 4 snapshots of my MD simulations that I tested (though, in all honesty, I still haven't had time to verify whether the values found by the three methodologies match, etc.).

Thanks again for your help, and best regards,

Stefano
________________________________
Da: Christopher Woods <notifications@github.com>
Inviato: martedì 28 gennaio 2020 09:33
A: michellab/Sire <Sire@noreply.github.com>
Cc: sas103 <stefanoas@hotmail.com>; Author <author@noreply.github.com>
Oggetto: Re: [michellab/Sire] Waterview has issues parsing files that are parsed without problems by waterswap (#306)


Hi Stefano. Thanks for raising this, and it is the best place. I’ll try to take a look this afternoon. I’m guessing there is some code rot as waterview is little used and not really marketted. The analytics show it has only ever been used twice in its history...!

Out of interest, what are you using waterview for, and how did you find it? Just curious as the only thing public about it is a talk I gave years ago and a YouTube video ;-)

**Comment by chryswoods** [comment]
The problem is that we didn't update WaterView.py when we slightly changed the search API to return lists instead of single molecules. I've committed a fix that will hopefully make things work for you (https://github.com/michellab/Sire/commit/52c51af874f114f34c77a157c3bd132503a42a23). I'm not able to check as your zipfile didn't include the topology file.

The CI/CD system for Sire should now be compiling and creating a new development binary. The easiest way for you to apply the fix is either

1. wait for this to complete and install the latest development binary, e.g. `conda install -c conda-forge -c omnia -c michellab/label/dev sire`.

2. or even easier, download the updated [WaterView.py](https://raw.githubusercontent.com/michellab/Sire/52c51af874f114f34c77a157c3bd132503a42a23/wrapper/Tools/WaterView.py) and copy this to overwrite the old version you have at `/home/serapian/sire.app/lib/python3.7/site-packages/Sire/Tools/WaterView.py`.

However... WaterView doesn't do what you think it does. WaterView will run Monte Carlo of the swapped water molecules to generate an average position map of where the water molecules would be placed if there was no ligand present. This is useful as, generally, a ligand displaces water, and so negative polar parts of the ligand should displace the oxygens of "semi-static" water molecules, while positive parts of the ligand should displace the hydrogens. [This video](https://www.youtube.com/watch?v=NMMTu6mut6I) shows the output of WaterView. The trajectory shows the motion of water molecules that displace the ligand, and the 3D density plots show the regions of occupancy of the waters. The most-occupied regions overlay very well with the polar parts of the ligand (e.g. the nitrogens). This explained why, in this case, the isomer of the ligand that had the nitrogens at those specific locations was effective, but isomers that had the nitrogens at different ring positions bound poorly. We never progressed this research so it was never written up (joys of being too busy a postdoc).

What I think you want is to plot the free energy decomposition for each water molecule and each residue from the waterswap calculation. This is output by the waterswap calculation itself, written into each iteration's output file. Simply read in the free energies from each output file, average them over the iterations you want (e.g. 400-1000) and then use them to colour the residues and water molecules in a program like VMD.

**Comment by sas103** [comment]
Hi Chris, and sorry for getting back to you so late.

First things first: I have done as you said, substituting the scripts, and the program has run farther than its ever done before. However, it now fails with another error: I am attaching another zip file, containing all the input files (this time the topology too), the slurm script used to launch, STERR and STDOUT.

Output files correctly produced include: complex.s3, waterbox.s3, stage1_fixed_atoms.pdb, stage1_mobile_atoms.pdb, stage2_fixed_atoms.pdb, stage2_mobile_atoms.pdb.

I have no idea whether the error is due to a problem with my system, to another bug, or to the HPC architecture. At any rate, I have noted what you said in your e-mail about differences between WaterSwap and WaterView, and indeed, as I had replied to you privately, it is correct to say that the information I actually want is the free energy decomposition for each water molecule (which is given by WaterSwap.

As such, the issue is not really as urgent. I am keeping the issue open, but we can close it if you see fit.

[files_to_reproduce_error_newWaterViewpy_script_2.zip](https://github.com/michellab/Sire/files/4276896/files_to_reproduce_error_newWaterViewpy_script_2.zip)

**Comment by chryswoods** [comment]
It looks like the system has tripped one of the consistency tests that are performed while the code is running. Monte Carlo updates the total energy by summing together small energy deltas from each move. This total can drift from the recalculated total if there are bugs in the program, numerical drift, or weird events (e.g. moves that cause large changes in energy that are accepted). In any of these cases, the calculation would be suspect. Sire (and thus waterview) recalculates the total energy periodically and compares against the running total. If the difference between the two is too large, then it halts on an error, printing out the energy components that have drifted. In this case, the drift is in the "mobile-fixed" energy, e.g.

```
The running total energy (-5205.58 kcal mol-1) disagrees with the total energy as calculated from scratch (-5204.46 kcal mol-1). Energy components which show disagreements greater than 0.1 kcal mol-1 are listed below:E_{total}: -5205.58 versus -5204.46 kcal mol-1. Difference equals 1.11724 kcal mol-1
E_{mobile-fixed}^{CLJ}: -3664.45 versus -3663.34 kcal mol-1. Difference equals 1.11724 kcal mol-1
E_{mobile-fixed}^{LJ}: 9.07623 versus 8.25538 kcal mol-1. Difference equals 0.820854 kcal mol-1
E_{mobile-fixed}^{coulomb}: -3673.53 versus -3671.59 kcal mol-1. Difference equals 1.93809 kcal mol-1
```

This is the energy between moving atoms and fixed atoms. I suspect this is a numerical drift caused by the mobile atoms being pushed into the fixed atoms because of too much pressure (too many water molecules have been substituted for the ligand). This happened because the algorithm to add water molecules is quite basic, and can fail for certain ligand shapes (e.g. long and thin). As waterview is not going to do what you want, and is not an advertised or used program, I don't think it is worth investigating further. As such, I will close this issue, and just say that waterview doesn't work well with your system.

**Comment by sas103** [comment]
Thanks for having a look at this, Chris!
